### PR TITLE
feat(metriken): rewrite metrics library

### PR DESCRIPTION
**Pull request by brayniac**
A clean reimplementation of the metriken metrics library. That makes some slightly different design choices. Metrics may now have associated metadata and an associated formatter. This allows for adding scoped metrics and to define a particular formatter for each metric.

Overall this adds a lot of flexibility through a richer metrics representation.

### CODE REVIEW

**Review by swlynch99 — CHANGES_REQUESTED** [review]
I haven't fully completed the review but I figure I'll leave some comments before I leave for today.

Some more specific comments on implementation details:
- I think either `name` and `description` should be moved out of the metadata.
- `name` should not be optional.
- `as_any` on the metric trait should return an `Option<&dyn Any>` (as it did before) to reflect whether the metric is active or not.
- `LazyCounter` and `LazyGauge` should be wrappers around `OnceLock<Counter>` and `OnceLock<Gauge>`, respectively, instead of copy-pasting the code.
- The `MetricEntry` trait is rather messy. Instead of returning an owned `HashMap` from `MetricEntry::metadata` it should be probably be something that wraps either the `phf::Map` or the `HashMap`.
- A whole bunch of things used to have docs but don't anymore.

This isn't exhaustive and I'll have to keep reviewing it tomorrow.

In general I don't have huge issues with the new features you've added. I have some smaller issues (as we've already discussed) but they are fixable. However, I feel like the larger rewrite has made the crate less consistent and harder to use. I think we should sit down and try and hash out the design instead of me nitpicking it over github comments.

File: metriken/src/metrics/mod.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+mod dynamic;
+mod r#static;
+
+pub use dynamic::*;
+pub use r#static::*;
+
+pub trait Metric: Send + Sync + 'static {
+    fn as_any(&self) -> &dyn Any;
+}
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/src/metrics/mod.rs:11]: I actually don't like the previous trait all that much. Let's discuss outside of the review.

File: metriken/derive/src/metric.rs [hunk 2]
```diff
@@ -110,71 +75,116 @@
...
+
+    let attrs: Vec<_> = metadata
+        .0
+        .into_iter()
+        .map(|(_, entry)| {
+            let key = entry.name.to_literal();
+            let value = entry.value;
+
+            quote!( #key => #value )
+        })
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/derive/src/metric.rs]: Fixed

File: metriken/derive/src/args.rs [hunk 3]
```diff
@@ -45,3 +64,155 @@
...
+
+    fn insert_or_duplicate(&mut self, arg: SingleArg<Self::Inner>) -> syn::Result<()>;
+}
+
+impl<T> SingleArgExt for Option<SingleArg<T>> {
+    type Inner = T;
+
+    fn insert_or_duplicate(&mut self, arg: SingleArg<Self::Inner>) -> syn::Result<()> {
+        match self {
+            None => Ok(*self = Some(arg)),
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/derive/src/args.rs]: Fixed

File: metriken/derive/src/metric.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/derive/src/metric.rs]: Fixed

File: metriken/src/lib.rs [hunk 4]
```diff
@@ -1,426 +1,295 @@
...
-                .maximum_value($max as _)
-                .min_resolution(1)
-                .min_resolution_range(1024)
-                .span($crate::export::Duration::<$crate::export::Nanoseconds<u64>>::from_secs(60))
-                .resolution($crate::export::Duration::<$crate::export::Nanoseconds<u64>>::from_secs(1))
-                .build()
-                .expect("bad heatmap configuration")
-        });
-    };
+pub fn deregister_all() {
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/src/lib.rs]: Removed

File: metriken/src/lib.rs [hunk 5]
```diff
@@ -1,426 +1,295 @@
...
-    ///
-    /// [`Any`]: std::any::Any
-    fn as_any(&self) -> Option<&dyn Any>;
+#[macro_export]
+#[rustfmt::skip]
+macro_rules! metadata {
+    ($($tts:tt)*) => {
+        metriken::Metadata::new(metriken::phf_map!($($tts)*))
+    };
 }
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/src/lib.rs]: Done.

File: metriken/src/metrics/mod.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+impl Metrics {
+    pub(crate) fn deregister_all() {
+        let metrics: Vec<Arc<dyn Metric>> =
+            metrics().dynamic.iter().map(|v| v.metric.clone()).collect();
+
+        for metric in metrics {
+            DYNAMIC_REGISTRY.deregister(metric)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/src/metrics/mod.rs]: This isn't needed. Removed.

File: metriken/src/counters/standard.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    fn as_any(&self) -> &(dyn std::any::Any + 'static) {
+        self
+    }
+}
+
+impl Metric for &'static Counter {
+    fn as_any(&self) -> &(dyn std::any::Any + 'static) {
+        self
+    }
+}
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/src/counters/standard.rs]: Done.

File: metriken/src/metrics/dynamic/registry.rs [hunk 8]
```diff
@@ -0,0 +1,36 @@
+use crate::*;
+
+pub(crate) struct DynamicRegistry {
+    metrics: RwLock<Vec<DynamicEntry>>,
```
[the remark this answers is not in the document]
Comment (brayniac) [reply metriken/src/metrics/dynamic/registry.rs]: Done.

**Comment by brayniac** [comment]
Closing this as the changes needed were made across several smaller PRs.

**Pull request closed without merging**